### PR TITLE
Adding self query for vectara

### PR DESCRIPTION
**Pull request by adeelehsan**
Add support for Self-query for the Vectara integration.
Vectara's [filtering](https://docs.vectara.com/docs/common-use-cases/filtering-by-metadata/filter-overview) is simply a string in pseudo-SQL format. Meta-data fields exists at the "doc" level and "part" level. Part refers to a piece of the document.
For self-querying we would like the filter to apply to the "doc" level metadata.

Fixes # ([2555](https://github.com/langchain-ai/langchainjs/issues/2555))

### CODE REVIEW

File: langchain/src/retrievers/self_query/vectara.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import { isFilterEmpty } from "./utils.js";
+
+type AllowedOperator = Exclude<Operator, NOT>;
+
+export type VectaraVisitorResult =
+  | VectaraOperationResult
+  | VectaraComparisonResult
+  | VectaraVisitorStructuredQueryResult;
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type VectaraOperationResult = String;
```
Comment (jacoblee93) [inline langchain/src/retrievers/self_query/vectara.ts:23]: This doesn't cause problems somewhere down the road?

**Comment by jacoblee93** [comment]
Thanks @adeelehsan! `String` type hack is a bit suspect but if it works I am happy to merge.

**Comment by adeelehsan** [comment]
> Thanks @adeelehsan! `String` type hack is a bit suspect but if it works I am happy to merge.

@jacoblee93 thanks for reviewing it. 
Yeah it works. `Vectara` is expecting filters as a string instead of object.
Default implementation of  methods `operation` and `comparison` methods are expecting return type object thats why I have defined the String as object.

  You can merge it.

**Comment by jacoblee93** [comment]
Thanks!

**Pull request merged**